restcomm/restcomm.mgcp/src/main/java/org/restcomm/connect/mgcp/MockPlayDelayMediaGateway.java

```diff
@@ -0,0 +1,724 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+package org.restcomm.connect.mgcp;
+
+import akka.actor.Actor;
+import akka.actor.ActorRef;
+import akka.actor.ActorSystem;
+import akka.actor.Props;
+import akka.actor.ReceiveTimeout;
+import akka.actor.UntypedActor;
+import akka.actor.UntypedActorContext;
+import akka.actor.UntypedActorFactory;
+import akka.event.Logging;
+import akka.event.LoggingAdapter;
+import jain.protocol.ip.mgcp.JainMgcpCommandEvent;
+import jain.protocol.ip.mgcp.JainMgcpResponseEvent;
+import jain.protocol.ip.mgcp.message.CreateConnectionResponse;
+import jain.protocol.ip.mgcp.message.DeleteConnection;
+import jain.protocol.ip.mgcp.message.DeleteConnectionResponse;
+import jain.protocol.ip.mgcp.message.ModifyConnection;
+import jain.protocol.ip.mgcp.message.ModifyConnectionResponse;
+import jain.protocol.ip.mgcp.message.NotificationRequest;
+import jain.protocol.ip.mgcp.message.NotificationRequestResponse;
+import jain.protocol.ip.mgcp.message.Notify;
+import jain.protocol.ip.mgcp.message.parms.ConnectionDescriptor;
+import jain.protocol.ip.mgcp.message.parms.ConnectionIdentifier;
+import jain.protocol.ip.mgcp.message.parms.EndpointIdentifier;
+import jain.protocol.ip.mgcp.message.parms.EventName;
+import jain.protocol.ip.mgcp.message.parms.NotifiedEntity;
+import jain.protocol.ip.mgcp.message.parms.ReturnCode;
+import jain.protocol.ip.mgcp.pkg.MgcpEvent;
+import org.joda.time.DateTime;
+import org.mobicents.protocols.mgcp.jain.pkg.AUMgcpEvent;
+import org.mobicents.protocols.mgcp.jain.pkg.AUPackage;
+import org.restcomm.connect.commons.faulttolerance.RestcommUntypedActor;
+import org.restcomm.connect.commons.util.RevolvingCounter;
+import org.restcomm.connect.commons.util.WavUtils;
+import org.restcomm.connect.mgcp.stats.MgcpConnectionAdded;
+import org.restcomm.connect.mgcp.stats.MgcpConnectionDeleted;
+import org.restcomm.connect.mgcp.stats.MgcpEndpointAdded;
+import org.restcomm.connect.mgcp.stats.MgcpEndpointDeleted;
+import scala.concurrent.duration.Duration;
+
+import javax.sdp.SdpFactory;
+import javax.sdp.SdpParseException;
+import javax.sdp.SessionDescription;
+import javax.sound.sampled.AudioFileFormat;
+import javax.sound.sampled.AudioFormat;
+import javax.sound.sampled.AudioInputStream;
+import javax.sound.sampled.AudioSystem;
+import javax.sound.sampled.UnsupportedAudioFileException;
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author hoan.h.luu@telestax.com
+ */
+public class MockPlayDelayMediaGateway extends RestcommUntypedActor {
```

Comment:
@xhoaluu a couple of clarifications needed:
1. as per issue description it does not seem that there is any issue related to play delay then why are we adding a  `MockPlayDelayMediaGateway`
2. If `MockPlayDelayMediaGateway` is needed for some reason, it should inherit the existing `MockMediaGateway` and enhance only the specific part of MediaGateway that we want to mock. can you please share which part of this MG is enhanced and do it by inheritance.

Reply:
@maria-farooq Thanks for comments

Issue is when Finish Dial verb having timeout(default is 30 seconds), If after Dial verb state of call is Playing, then timeout of dial verb will finish the call.

1/ As MediaMockGateway immediately send back NTFY after RQNT for playing an audio file, that mean I cannot create playing state for a while to wait Dial verb timeout occured. Then I need to delay 2 seconds between a RQNT and NTFY when RC request RMS playing cache file to make issue happened.

2/ I also want to inherit MockMediaGateway. but there are a lot of private functions to come to my edited code. For more easy to later modification, I copied whole MediaMockGateway to new class and added my modification.

https://github.com/RestComm/Restcomm-Connect/blob/aaa40256fd3451a01ef41edd11d6d5bb314d1a85/restcomm/restcomm.mgcp/src/main/java/org/restcomm/connect/mgcp/MockPlayDelayMediaGateway.java#L663

Reply:
@xhoaluu you can change `notify` method from `private` to `protected` and then override it in new class

Reply:
As I don't want to touch to legacy code, but If you said that, I can update it and avoid code duplicated.